drivers: sensor: microchip: mtch9010 Added MTCH9010 Device Support

robertperkel opened the pull request:
Added new driver for MTCH9010 liquid leak detector along with test patterns.

robertperkel committed:
drivers: sensor: microchip: mtch9010 Added mtch9010 support

Added support for MTCH9010 and applied edits from reviewers.

Signed-off-by: Robert Perkel <robert.perkel@microchip.com>

msalau commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+
+int mtch9010_decode_char_buffer(const char *buffer, enum mtch9010_output_format format,
+				struct mtch9010_result_t *result)
+{
```
`buffer` may not be null-terminated. I think it will be much safer to also pass its length to the function or explicitly add null-termination in `mtch9010_timeout_receive()`

robertperkel replied:
Added null-termination

kartben commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+		val->val1 = data->heartbeatErrorState;
+		break;
+	}
+	default: {
```
you're missing `SENSOR_CHAN_MTCH9010_MEAS_RESULT_AND_DELTA` case?

kartben commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+		.reference = MTCH9010_REF_VAL_INIT(inst),                                          \
+		.threshold = DT_INST_PROP(inst, mtch9010_detect_value),                            \
+		.heartbeatErrorState = false,                                                      \
+		.lastResult = {0, 0}};                                                             \
```
lastWake also needs to be initialized to zero to not have issues w/ first sample reading

kartben commented on drivers/sensor/microchip/mtch9010/Kconfig:
```diff
@@ -0,0 +1,102 @@
+		default 1
+		range 1 64
+		help
+		If the MTCH9010 is using the current measurement as it's reference, then this
```
```suggestion
		If the MTCH9010 is using the current measurement as its reference, then this
```

kartben commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+		mtch9010_command_send(uartDev, config, tBuffer);
+
+#ifdef CONFIG_MTCH9010_LOCK_ON_STARTUP
+		mtch9010_lock_device(dev);
```
undeclared function? did you mean mtch9010_lock_settings?

kartben commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+
+	/* SYSTEM LOCK */
+	if (gpio_is_ready_dt(&config->lockLine)) {
+		/* Keep LOW until ready to configure */
+		gpio_pin_configure_dt(&config->lockLine, GPIO_OUTPUT | GPIO_OUTPUT_INIT_HIGH);
```
comment doesn't match the code

robertperkel replied:
Comment is incorrect - fixed

msalau commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+		if (code == 0) {
+			count++;
+		}
+	}
```
Do responses end with `\r\n`? If yes - we can terminate reception when `\r\n` is received this making reception faster.

msalau commented on drivers/sensor/microchip/CMakeLists.txt:
```diff
@@ -1,7 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # zephyr-keep-sorted-start
+add_subdirectory_ifdef(CONFIG_DT_HAS_MICROCHIP_MTCH9010_ENABLED mtch9010)
```
```diff
-add_subdirectory_ifdef(CONFIG_DT_HAS_MICROCHIP_MTCH9010_ENABLED mtch9010)
+add_subdirectory_ifdef(CONFIG_MTCH9010 mtch9010)
```

robertperkel commented on tests/drivers/sensor/mtch9010/test_configs/data_format/current.overlay:
```diff
@@ -0,0 +1,16 @@
+&uart1 {
+    status = "okay";
+    current-speed = <36400>;
```
Good catch!

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+	/* CFG EN */
+	if (gpio_is_ready_dt(&config->enableCFGLine)) {
+		if (config->extendedMode) {
+			gpio_pin_set_dt(&config->enableUARTLine, GPIO_OUTPUT_LOW);
+		} else {
+			gpio_pin_set_dt(&config->enableUARTLine, GPIO_OUTPUT_HIGH);
```
Ah, I see. UART EN is active LOW

kartben replied:
no the issue is basically the same thing @msalau commented on. You want to be controlling enableCFGLine here, not enableUARTLine??

robertperkel replied:
Thanks - I think I see the issue now

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,792 @@
+};
+
+/* Device Init Macros */
+#define MTCH9010_OPERATING_MODE_INIT(inst)                                                         \
+	COND_CODE_1(DT_INST_ENUM_HAS_VALUE(inst, mtch9010_operating_mode,\
+		capacitive), (MTCH9010_CONDUCTIVE), (MTCH9010_CAPACITIVE))
```
I seem to remember a test error before I swapped it ... I'll look at it again

kartben commented:
> @kartben I'm looking at the twister failures, and it doesn't seem related to my PR, as I don't utilize with the file system. Can you confirm or rerun/ignore that twister error?

yep, had nothing to do with you, and fixed now :)

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+				   struct mtch9010_data_t *data)
+{
+	/* Note: nRESET is handled in device reset function */
+	int rtn = 0;
```
I'm not following - it looks like anything optional would be set to NULL and skipped

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+#ifdef CONFIG_MTCH9010_LOCK_ON_STARTUP
+	const struct mtch9010_config_t *config = dev->config;
+
+	if (!gpio_is_ready_dt(&config->lock_gpio)) {
```
Swapped to checking if the port is NULL, instead of gpio is ready. Lock can only be asserted after UART init

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+		}
+
+		/* Sleep Mode */
+		snprintf(temp_buffer, MTCH9010_INTERNAL_BUFFER_SIZE, "%d", config->sleep_time);
```
Good suggestion

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+				 const struct mtch9010_config_t *config)
+{
+#ifndef CONFIG_MTCH9010_RESET_ON_STARTUP
+	LOG_INST_WRN(config->log, "MTCH9010 reset on startup is disabled.");
```
I'm not opposed to this change, but I thought it was frowned upon to have a function that is not always compiled into the project.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+	}
+
+	/* Verify the \r\n is at the end */
+	if ((*endOfString == '\n') && (*(endOfString + 1) == '\r')) {
```
I double checked on my hardware yesterday, and it is /n/r. I will update the comments accordingly

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+
+/* Private Function Declarations */
+static int mtch9010_init(const struct device *dev);
+static int mtch9010_command_send(const struct device *dev, const struct mtch9010_config_t *config,
```
I like this idea

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010_priv.h:
```diff
@@ -0,0 +1,90 @@
+	/* Initialization mode of the reference */
+	enum mtch9010_reference_value_init ref_mode;
+	/* Logging Instance */
+	LOG_INSTANCE_PTR_DECLARE(log);
```
And since this is a leak detection ASIC, there is a decent change someone may want more than one at a time.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+	data->last_heartbeat = k_uptime_get();
+
+	/* Wait for boot-up */
+	k_msleep(CONFIG_MTCH9010_BOOT_TIME_MS);
```
Unfortunately, things stop working when I remove this delay. I have some theories, but for now, I'd like to keep this.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+		char temp_buffer[MTCH9010_INTERNAL_BUFFER_SIZE];
+
+		if (config->mode == MTCH9010_CAPACITIVE) {
+			mtch9010_command_send(uart_dev, config, "0");
```
Done - I left the enum -> int mapped ones alone, since those are more compact than a switch statement

robertperkel commented:
Looks like everything is good to go on our side 👍 Local test bench is functioning as expected

robertperkel commented:
Hi @msalau @kartben ,
Since the feature freeze date for v4.2 is this Friday, could you let me know any remaining changes and/or approve the PR? 

Additionally - do I need to notify anyone of this for v4.2, or is it automatic? 

Thanks for your feedback! 
Robert

kartben commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,861 @@
+		.uart_init = DT_PROP(DT_DRV_INST(inst), mtch9010_uart_config_enable),              \
+		.uart_dev = DEVICE_DT_GET_OR_NULL(DT_PARENT(DT_DRV_INST(inst))),                   \
+		.reset_gpio = GPIO_DT_SPEC_GET_OR(DT_DRV_INST(inst), mtch9010_reset_gpios, {0}),   \
+		.mode_gpio = GPIO_DT_SPEC_GET_OR(DT_DRV_INST(inst), mtch9010_mode_gpios, {0}),     \
```
you're not doing anything with this? It's not even defined in the binding. Would need to drop the associated tests too...

robertperkel replied:
Which device? I see config enable being referenced. As for the other ones - this sensor only can check mode on startup, and only in certain configurations (1). After POR for the sensor, the values are polled and locked in until the next power cycle. 

Notes
(1) - If not in UART config mode, then it will poll the GPIO. If not in UART config, then this will do nothing.

kartben replied:
What I mean is the binding doesn't have any `mode-gpios` property so I am not sure what this is trying to do here

robertperkel replied:
You're correct, it's never referenced. I'll add that to init GPIO

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,861 @@
+		.last_result = {0, 0, 0},                                                          \
+	};                                                                                         \
+	static const struct mtch9010_config_t mtch9010_config_##inst = {                           \
+		.uart_init = DT_PROP(DT_DRV_INST(inst), mtch9010_uart_config_enable),              \
+		.uart_dev = DEVICE_DT_GET_OR_NULL(DT_PARENT(DT_DRV_INST(inst))),                   \
```
This is where it gets tricky - uart_init refers to configuring the device over UART, but you could init the device through I/O and still get a UART output.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,861 @@
+			return -EINVAL;
+		}
+
+		result->delta = (uint16_t)temp;
```
Yes it is - Added test coverage to further verify delta can be negative

kartben commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,873 @@
+		LOG_INST_DBG(config->log, "UART setup not enabled");
+
+#ifdef CONFIG_MTCH9010_LOCK_ON_STARTUP
+		BUILD_ASSERT(true, "MTCH9010_LOCK_ON_STARTUP cannot be set without UART enabled");
```
Please do not resolve comments. This is for the reviewers to do

kartben replied:
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

robertperkel replied:
The BUILD_ASSERT keeps causing issues, and since this is an optional feature anyway, I've removed the BUILD_ASSERT.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,861 @@
+		char temp_buffer[MTCH9010_INTERNAL_BUFFER_SIZE];
+
+		if (config->mode == MTCH9010_CAPACITIVE) {
+			mtch9010_command_send(dev, MTCH9010_CMD_STR_CAPACITIVE_MODE);
```
The reason I had this was due to earlier issues with RX failing, but actually passing. I'm modifying the code to return -EIO if these fail.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,861 @@
+
+		for (uint8_t i = 0; i < MTCH9010_INTERNAL_BUFFER_SIZE; i++) {
+			temp_buffer[i] = '\0';
+		}
```
Removed function

robertperkel commented:
Fixed some lingering issues noted by reviewers:
- Fixed a few instances of camelCase that were missed
- Modified init behavior to return error if sending a command failed.
- Return error if attempting to fetch from uart without a valid uart instance

robertperkel commented on drivers/sensor/microchip/mtch9010/Kconfig:
```diff
@@ -0,0 +1,95 @@
+		  If set, this option will assert SYS_LK after initialization.
+		  The sensor will use these parameters if SYS_LK is asserted on the next POR.
+
+	config MTCH9010_RESET_TIME_MS
```
Fair enough - the only reason why someone might want to configure this is if we release a part with improved characteristics in the future

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,861 @@
+	if (gpio_is_ready_dt(&config->heartbeat_gpio)) {
+		gpio_pin_configure_dt(&config->heartbeat_gpio, GPIO_INPUT);
+#ifdef CONFIG_MTCH9010_HEARTBEAT_MONITORING_ENABLE
+		rtn = gpio_pin_interrupt_configure_dt(&config->heartbeat_gpio,
+						      GPIO_INT_EDGE_RISING);
+		if (rtn == 0) {
+			gpio_init_callback(&data->heartbeat_cb, mtch9010_heartbeat_callback,
+					   BIT(config->heartbeat_gpio.pin));
+			rtn = gpio_add_callback_dt(&config->heartbeat_gpio, &data->heartbeat_cb);
+			if (rtn == 0) {
+				rtn = gpio_pin_interrupt_configure_dt(&config->heartbeat_gpio,
+								      GPIO_INT_EDGE_BOTH);
```
I'm a bit confused as well. 😆 

I think I was debating between two different ways to implement this functionality, and I ended up with a hybrid. Removed first instance of configure as fix

robertperkel commented:
Removed BUILD_ASSERT due to persistent issues with logic. (System Lock does not affect core functionality - is optional)

robertperkel commented:
Added __ASSERT Validation of UART Parameters to Init function 
Made UART Mandatory

robertperkel commented:
@kartben @msalau 
Any other changes to make to this PR?

robertperkel commented:
Also, should this be retagged to v4.3.0 milestone, or is v4.2.0 still open?

kartben commented:
> Also, should this be retagged to v4.3.0 milestone, or is v4.2.0 still open?

Yes, unfortunately this missed the deadline for the 4.2 release as feature freeze was last Friday. Getting a brand new 2000-line PR reviewed in just 10 short days right as everyone was wrapping up their own contributions was probably a bit ambitious unfortunately. Hope you understand!

robertperkel commented:
Just to check - until v4.2 is released (7/18), merging is blocked, correct?

kartben commented:
> Just to check - until v4.2 is released (7/18), merging is blocked, correct?
> 

Correct. Everything goes into `main` branch so until vN is released, vN+1 features are queuing up :)

robertperkel commented:
Since 4.2 has released, can the reviewers take a look at this when they get a moment?

We're planning a follow-up to this merge with another commit containing a device sample

JarmouniA reviewed (changes requested):
Put tests-related changes in a separate commit.

robertperkel commented:
Removed mtch9010 tags from .yaml testing files

robertperkel commented on include/zephyr/drivers/sensor/mtch9010.h:
The sensor_channel_mtch9010 enumeration adds to the SENSOR_CHAN enumeration per the specification. The other structs/enums can be moved to the private .h file.

robertperkel commented on drivers/sensor/microchip/mtch9010/mtch9010.c:
```diff
@@ -0,0 +1,930 @@
+/* Device Init Macros */
+#define MTCH9010_OPERATING_MODE_INIT(inst)                                                         \
+	COND_CODE_1(DT_INST_ENUM_HAS_VALUE(inst, mtch9010_operating_mode,\
+		mtch9010_mode_capacitive), (MTCH9010_CAPACITIVE), (MTCH9010_CONDUCTIVE))
```
Thanks for the suggestion - what I did was to remove the dt-bindings header I made (from an earlier suggestion) and convert my COND_CODE lookups to DT_STRING_UPPER_TOKEN, which maps to the enumerations already defined.

robertperkel replied:
Won't that cause issues with the enumeration definitions?

JarmouniA reviewed (commented):
Still missing the split into 2 commits.

robertperkel committed:
tests: drivers: sensor: Added MTCH9010 Tests

Added testsuite to verify MTCH9010 sensor driver

Signed-off-by: Robert Perkel <robert.perkel@microchip.com>

robertperkel commented on dts/bindings/sensor/microchip,mtch9010.yaml:
```diff
@@ -0,0 +1,127 @@
+include: [sensor-device.yaml, uart-device.yaml]
+
+properties:
+  mtch9010-operating-mode:
```
I think I split it correctly
